include/eosio/vm/host_function.hpp

```diff
@@ -17,522 +21,371 @@
-      // returns a value_getter or void if it didn't match.
-      template<typename T, typename Cons, std::size_t... Is>
-      auto try_value_getter(std::index_sequence<Is...>)
-         -> value_getter<sizeof...(Is), decltype(match_from_wasm<cons_item_t<Is, Cons>...>::apply(&wasm_type_converter<T>::from_wasm))>;
-      // Fallback
-      template<typename T, typename Cons>
-      auto try_value_getter(...) -> void;
-
-      template<std::size_t N, typename T, typename Cons>
-      using try_value_getter_t = decltype(try_value_getter<T, Cons>(std::make_index_sequence<N>{}));
-
-      // Error type to hopefully make a somewhat understandable error message when
-      // the user did not provide a correct overload of from_wasm.
-      template<typename T, typename... Tail>
-      struct no_viable_overload_of_from_wasm {};
-
-      // Searches for a match of from_wasm following the principal of maximum munch.
-      template<std::size_t N, typename T, typename Tail>
-      constexpr auto make_value_getter_impl() {
-         if constexpr(std::is_same_v<try_value_getter_t<N, T, Tail>, void>) {
-            if constexpr (N == 0) {
-               return no_viable_overload_of_from_wasm<T>{};
-            } else {
-               return make_value_getter_impl<N-1, T, Tail>();
-            }
+      template <typename S, typename Type_Converter>
+      inline constexpr std::size_t skip_amount() {
+         if constexpr (has_as_value<S, Type_Converter>()) {
+            return 1;
          } else {
-            return try_value_getter_t<N, T, Tail>{};
+            return std::tuple_size_v<from_wasm_type_deducer_t<Type_Converter, S>>;
          }
       }
 
-      template<int N>
-      struct fold_cons_impl;
-      template<typename... T>
-      using fold_cons = typename fold_cons_impl<sizeof...(T)>::template fn<T...>;
-      template<>
-      struct fold_cons_impl<0> {
-         template<typename... T>
-         using fn = nil_t;
-      };
-      template<int N>
-      struct fold_cons_impl {
-         template<typename T0, typename... T>
-         using fn = cons<cons_item<T0, T0>, fold_cons<T...>>;
-      };
-
-      template<typename T, typename Tail>
-      constexpr auto make_value_getter() {
-         return make_value_getter_impl<Tail::size, T, Tail>();
+      template <typename Args, std::size_t At, std::size_t Skip_Amt, class Type_Converter>
+      inline constexpr auto get_values(Type_Converter& tc) {
+         if constexpr (At >= std::tuple_size_v<Args>)
+            return std::tuple<>{};
+         else {
+            using source_t = std::tuple_element_t<At, Args>;
+            constexpr std::size_t skip_amt = skip_amount<source_t, Type_Converter>();
+            using converted_t = decltype(create_value<Args, source_t, Skip_Amt>(tc, std::make_index_sequence<skip_amt>{}));
+            using wrapped_t = std::conditional_t<std::is_reference_v<converted_t>, reference<std::remove_reference_t<converted_t>>, converted_t>;
```

Comment:
I would really love it if we could just change this into a `static_assert` that disallows references in the host function arguments (or at the very least non-const references). Then we could get rid of `reference`, `is_reference_type`, `is_reference_type_v`, and `detail::convert_type`.

I don't think it is a good idea generally (not just for EOSIO) for the host functions to provide reference types in the argument list. If they want an array of items (either `const` or not) with a run-time specified size, they should just use a `span`. If they want to be able to modify a value passed in by the WASM, they should use a pointer. They can also use a pointer to a `const` value, if they want to be able to efficiently read a large structure from WASM linear memory. But in no case is a reference necessary. What do you think @larryk85?

Reply:
I don't understand why reference arguments are a problem.  I don't agree with forbidding them, but that doesn't mean that we need to support them either.  Reference arguments already require a custom `from_wasm`, so there's really no need to have `reference` built in.

Reply:
Okay. I'm fine with not forbidding if we don't need to have special case handling of references in the base.